server/api/auth/linkedin.js

```diff
@@ -0,0 +1,110 @@
+
+if (useDevApiServer) {
+  callbackURL = `http://localhost:${PORT}/api/auth/linkedin/callback`;
+} else {
+  callbackURL = `${rootUrl}/api/auth/linkedin/callback`;
+}
+
+const strategyOptions = {
+  clientID,
+  clientSecret,
+  callbackURL,
+  scope: ['r_emailaddress', 'r_liteprofile'],
+  passReqToCallback: true,
+};
+
+const verifyCallback = (req, accessToken, refreshToken, rawReturn, profile, done) => {
+  // We can can use util function to generate id token to match OIDC so that we can use
+  // our custom id provider in Flex
+
+  console.log('Profile', profile);
+
+  const locale = Object.keys(profile._json.firstName.localized)[0];
+
+  const firstName = profile._json.firstName.localized[locale];
+  const lastName = profile._json.lastName.localized[locale];
+  const email = profile.emails[0].value;
+
+  // LikedIn API doesn't return information if the email is verified or not directly.
+  // However, it seems that with OAUTH2 flow authentication is not possible if the email is not verified.
+  // There is no official documentation about this, but through testing it seems like this can be trusted
+  // For reference: https://stackoverflow.com/questions/19278201/oauth-request-verified-email-address-from-linkedin
+
+  const user = {
+    userId: profile.id,
+    profile: {
+      firstName,
+      lastName,
+      email,
+      emailVerified: true,
+    },
+  };
+
+  const state = req.query.state;
+  const queryParams = JSON.parse(state);
+
+  const { from, defaultReturn, defaultConfirm } = queryParams;
+
+  const idpToken = createIdToken(user)
+    .then(idpToken => {
+      const userData = {
+        email,
+        firstName,
+        lastName,
+        idpToken,
+        from,
+        defaultReturn,
+        defaultConfirm,
+      };
+
+      console.log('userData:', userData);
+
+      done(null, userData);
+    })
+    .catch(e => console.error(e));
+};
+
+// ClientId is required when adding a new Linkedin strategy to passport
+if (clientID) {
+  passport.use(new LinkedInStrategy(strategyOptions, verifyCallback));
+}
+
+exports.authenticateLinkedin = (req, res, next) => {
+  const from = req.query.from ? req.query.from : null;
+  const defaultReturn = req.query.defaultReturn ? req.query.defaultReturn : null;
+  const defaultConfirm = req.query.defaultConfirm ? req.query.defaultConfirm : null;
+
+  const params = {
+    ...(!!from && { from }),
+    ...(!!defaultReturn && { defaultReturn }),
+    ...(!!defaultConfirm && { defaultConfirm }),
+  };
+
+  const paramsAsString = JSON.stringify(params);
+
+  passport.authenticate('linkedin', {
+    state: paramsAsString,
+  })(req, res, next);
+};
+
+// Use custom callback for calling loginWithIdp enpoint
+// to log in the user to Flex with the data from Linkedin
+exports.authenticateLinkedinCallback = (req, res, next) => {
+  passport.authenticate('linkedin', function(err, user) {
+    loginWithIdp(err, user, req, res, 'client-id', 'linkedin');
```

Comment:
I'm quite confused about how this 'client-id' string works.
In loginWithIdp.js, there's **CLIENT_ID** constant and **clientId** param and they are not the same. There probably should be some kind of renaming made to these client id variables and values.

Reply:
My initial idea was that we are not going to keep the LinkedIn implementation in FTW so I cut some corners when I did testing. So this is one of those parts. But if we want to keep this in the repo, then this of course needs to be fixed. 

Anyway, this 'client-id' is just a hard-coded test value. This 'client-id' the same thing you should have in process.env.CUSTOM_OIDC_CLIENT_ID and it's the value that should match the client id you have in Console. Most likely that environment variable is named to match the identity provider you are using (LinkedIn, Instagram, your own system etc.).

Reply:
Makes sense 👍 

How about the renaming of things in loginWithIdp.js file? (And sorry, this should have been addressed already in previous PRs.) In that file, there is CLIENT_ID referring to Flex client id and "clientId" referring to IDP client id - and then code like `clientId: CLIENT_ID,` - so, I'd imagine that this could be pretty confusing for someone who tries to understand what happens under the hood.

Reply:
This is a really good point and I think renaming that parameter makes sense. I added that change to this PR now.